### PR TITLE
restructuring towards python client library #455

### PR DESCRIPTION
**Pull request by bzeiss**
So this is my restructuring attempt and starting a common python client library. I think this should be very near the state that it should be. 

A summary of what when on here:

- I have moved file_cmds, ractl_cmds and socket_cmds into a new python module common/src/rascsi
- ractl_cmds, socket_cmds and file_cmds have been moved into python classes and parameterized with respect to token, locale, rascsi host and port.
- Everything in common/src/rascsi has been de-flasked, i.e., all web specific stuff has been moved out of the common python module and moved into web. This is reflected on the one hand in changes in the return payloads of file_cmds and ractl_cmds so that there isn't any content that needs translation in there. Instead, those strings have been replaced by return codes in return_codes.py. In the web module, the return codes are mapped back to translated strings using the generic class return_code_mapper.py.
- Web specific error handling has also been removed from those common classes. This yields a subclass socket_cmds_flask.py for usage in web.py instead of the direct usage of socket_cmds.py in common. In order to make this generic, a bunch of custom exceptions have been defined in exceptions.py that are passed to the child class to interprete those specific error situations.
- start.sh scripts have been adapted to find the common python rascsi module
- easy_install.sh has been adapted to generate the protobuf python interface only once for the common module. 
- web and oled modules have been adapted use use the new client library. 
  - web/return_code_mapper.py takes care of mapping return_codes + parameters back to i18n strings.
  - I have removed code in oled that was now duplicated with common/src/racscsi (and previously web).
  - Due to the usage of the common python lib, I have added the now cheap added feature to direct the web and oled app to a different rascsi host/port.

I have tested the migration from the latest release to this branch and to me it seems to work fine. However, I wasn't able easily to trigger all situations that are covered in the return_code_mapper.py. The following strings are still open for test as of writing:

        ReturnCodes.RENAMEFILE_UNABLE_TO_MOVE: _(u"Unable to move file to: %(target_path)s"),
        ReturnCodes.WRITECONFIG_COULD_NOT_WRITE: _(u"Could not write to file: %(file_name)s"),
        ReturnCodes.READCONFIG_COULD_NOT_READ: _(u"Could not read configuration file: %(file_name)s"),
        ReturnCodes.WRITEDRIVEPROPS_COULD_NOT_WRITE: _(u"Could not write to properties file: %(file_path)s"),
        ReturnCodes.READDRIVEPROPS_SUCCESS: _(u"Read properties from file: %(file_path)s"),
        ReturnCodes.READDRIVEPROPS_COULD_NOT_READ: _(u"Could not read properties from file: %(file_path)s"),

I'm happy to hear the feedback. It would definately be good if a few people could give this a test run before this is merged.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -5,11 +5,10 @@ core
...
-python/web/src/rascsi_interface_pb2.py
-python/oled/current
-python/oled/src/rascsi_interface_pb2.py
+current
+rascsi_interface_pb2.py
 src/raspberrypi/hfdisk/
 *~
 messages.pot
 messages.mo
+s.sh
```
Comment (rdmark) [inline .gitignore:14]: What is s.sh?

**Comment by rdmark** [comment]
@bzeiss I think this is turning out well overall! Thanks again for your effort.

Did you get the chance to run pylint against each of the source files? I ran it on the major files and spotted a handful of potential code quality issues, plus some newly introduced noise (e.g. a whole lot of "Method could be a function (no-self-use)").
Would you prefer to address each yourself, or should I point out the ones that I think should be addressed?
There is a bunch of "known issues" that are definitely out of scope, so I'm not expecting a 100% quality score. :)

And while on the topic of pylint, two structural things:
I think it'd be great to have pylinting documented in a README under the top level python/ dir. Right now, there's a paragraph in the web/README.md file that could be cut out and made into its own README.
Secondly, what to do about the .pylintrc file? Before, I had a symlink in each python source dir so that pylint could find it automatically. Should we replicate this in the new structure, or do you have a better suggestion?

File: .gitignore (hunk 1 above)
Comment (bzeiss) [reply .gitignore:14]: Just a custom temporary start script that I use to mess around with stuff or test specific stuff that shouldn't be comitted. I can remove it from the .gitignore if you don't like it. It just means I have to take more care then what I commit and push.

**Comment by bzeiss** [comment]
Thanks!

True, pylint I didn't do, but should've done. I mostly rely on the IDEA code inspection. Don't worry about commenting on each issue, that is too much unnecessary work for you. I will run it and try to improve and what is complained about.

Yes, I'm aware about the pylint parts of the Web README. I will take care of this and take a look what a good solution for the .pylintrc file is. I have to play around with it to make a suggestion.

**Comment by rdmark** [comment]
@bzeiss Sounds good! And please don't hesitate to propose changing the linting rules, if it's complaining loudly about non-issues.

**Comment by bzeiss** [comment]
@rdmark I have pushed a new commit to this branch containing a lot of little pylint fixes. Actually, most fixes weren't related to the restructuring ;-)

Anyway, there are still a lot of open issues, like too broad exceptions, suggested f-string usage etc.. You will see that I have also added a few ignore annotations to shut pylint up for specific things:
- pylint wasn't able to properly figure out the import paths. I have tried to fiddle around with the sys.path in the init-hook= line in .pylintrc, but didn't manage to make it properly recognize the imports. Not sure what is happening here, but I thought it is probably better to locally disable that check than to disable it altogether. I may give this another try to properly make it find the packages.
- At some places, pylint wanted me to make functions static. As with the IDEA code inspections, I have disabled those as well as I think those classes should be instantiated.

Also, I have added a README.md and a .pylintrc. I think the config file is fine, I rather prefer to disable checks locally (and thus documenting an actual developer decision to purposefully accept something to be the way it is) than to disable rules altogether.

Let me know whether you like the new changes.

**Comment by bzeiss** [comment]
@rdmark One more thing: pylint suggested to provide the encoding for open and save explicitly. For compatibility, I have used ISO-8859-1 and not UTF-8. I wasn't actually too sure about this, but I thought that utf-8 based filenames are probably still rather rare so I think it is probably sufficient to save in ISO-8859-1.

File: .gitignore (hunk 1 above)
Comment (rdmark) [reply .gitignore:14]: The way I think about it is that .gitignore should be readable so that future collaborators can make informed decisions when updating it. How about adding a comment preceding that line, f.e.
```
# Temporary user files
```
Actually 'current' kind of falls into this category too, although its creation is scripted.

**Comment by rdmark** [comment]
@bzeiss The import path not being recognized is a problem, since it may mask other real code quality issues. If we can't figure it out here, an issue ticket is appropriate, I think.

File: python/common/src/rascsi/socket_cmds.py [hunk 2]
```diff
@@ -79,7 +85,8 @@ def send_over_socket(self, sock, payload):
             response_message = b''.join(chunks)
             return response_message
 
-        error_message = "The response from RaSCSI did not contain a protobuf header. RaSCSI may have crashed."
+        error_message = "The response from RaSCSI did not contain a protobuf header. " \
+                        "RaSCSI may have crashed."
```
Comment (rdmark) [inline python/common/src/rascsi/socket_cmds.py:89]: Nitpick: I prefer the parenthesis syntax for string continuation, which is what I've tried to use consistently elsewhere in the codebase. It is also the [PEP-8 recommendation](https://www.python.org/dev/peps/pep-0008/#maximum-line-length).

> The preferred way of wrapping long lines is by using Python's implied line continuation inside parentheses, brackets and braces. Long lines can be broken over multiple lines by wrapping expressions in parentheses. These should be used in preference to using a backslash for line continuation.

File: python/web/src/web.py [hunk 3]
```diff
@@ -101,8 +103,8 @@ def index():
         abort(
             403,
             _(
-                u"RaSCSI is password protected. "
-                u"Start the Web Interface with the --password parameter."
+                "RaSCSI is password protected. "
+                "Start the Web Interface with the --password parameter."
```
Comment (rdmark) [inline python/web/src/web.py:107]: Was this and below changes due to a pylint rule? I may have been under the mistaken assumption that gettext wanted unicode strings. :)

**Comment by bzeiss** [comment]
> @bzeiss The import path not being recognized is a problem, since it may mask other real code quality issues. If we can't figure it out here, an issue ticket is appropriate, I think.

Let me try again tonight. I'm pretty sure there are no problems due to the import as the idea code inspection would complain about it. So there are two safety nets in my case, but it's true that others might rely on pylint alone, so it's better to have this working.

File: python/web/src/web.py (hunk 3 above)
Comment (bzeiss) [reply python/web/src/web.py:107]: Yea, it complained that it's superfluous as it's default with Python 3.
image

File: python/common/src/rascsi/socket_cmds.py (hunk 2 above)
Comment (bzeiss) [reply python/common/src/rascsi/socket_cmds.py:89]: Interesting. The IDEA code inspection is pretty nitpicky itself about the PEP recommendations, but the line break it performs itself seem to be wrong. I'll correct those.

File: .gitignore (hunk 1 above)
Comment (bzeiss) [reply .gitignore:14]: sure, I have added that comment. I suggest to not move current to after that comment as current is generated by the official start.sh script.

**Comment by bzeiss** [comment]
Branch updated as suggested and discussed on discord.

File: python/common/src/README.md [hunk 4]
```diff
@@ -42,3 +42,14 @@ sudo pip install pylint-venv
...
+Examples:
+```
+# check a single file
+pylint web/src/web.py
+
+# check the python modules
+pylint common/src
+pylint web/src
+pylint oled/src
+```
```
Comment (rdmark) [inline python/common/src/README.md:55]: The point I tried to make before was: Instead of having the pylint information (including the paragraphs above) repeat in every package's README, why don't we break this out into a separate python/README.md file? It reduces documentation duplication, and also makes it clearer from which dir to run the examples. What do you think?

**Review by rdmark — APPROVED** [review]
